iceoryx_posh/test/integrationtests/test_service_discovery.cpp

```diff
@@ -473,31 +485,126 @@ TEST_F(ServiceDiscovery_test, FindServiceReturnsMaxServices)
 
     auto serviceContainer = sut.findService(SERVICE, iox::capro::Wildcard, iox::capro::Wildcard);
 
-    ASSERT_FALSE(serviceContainer.has_error());
-    EXPECT_THAT(serviceContainer.value().size(), Eq(iox::MAX_NUMBER_OF_SERVICES));
-    EXPECT_TRUE(serviceContainer.value() == serviceContainerExp);
+
+    EXPECT_THAT(serviceContainer.size(), Eq(iox::MAX_NUMBER_OF_SERVICES));
+    EXPECT_TRUE(serviceContainer == serviceContainerExp);
 }
 
-/// @todo #415 #1074 this test is affected by the limits we set for service discovery,
-/// if the container capacity is larger than what we can send with sockets
-/// we will never cause an overflow
-TEST_F(ServiceDiscovery_test, FindServiceReturnsContainerOverflowErrorWhenMoreThanMaxServicesAreFound)
+TEST_F(ServiceDiscovery_test, ServiceDiscoveryIsAttachableToWaitSet)
 {
-    ::testing::Test::RecordProperty("TEST_ID", "f2f8d8c0-8712-4e7a-9e33-2b2a918f8a71");
-    const IdString_t SERVICE = "s";
-    const size_t numberOfInstances = (iox::MAX_NUMBER_OF_SERVICES + 1);
-    iox::cxx::vector<iox::popo::UntypedPublisher, numberOfInstances> publishers;
-    for (size_t i = 0; i < numberOfInstances; i++)
+    ::testing::Test::RecordProperty("TEST_ID", "fc0eeb7a-6f2a-481f-ae8a-1e17460e261f");
+    iox::popo::WaitSet<10U> waitSet;
+
+    waitSet
+        .attachEvent(sut,
+                     ServiceDiscoveryEvent::SERVICE_REGISTRY_HAS_CHANGED,
+                     0U,
+                     iox::popo::createNotificationCallback(testCallback))
+        .and_then([]() { GTEST_SUCCEED(); })
+        .or_else([](auto) { GTEST_FAIL() << "Could not attach to wait set"; });
+}
+
+TEST_F(ServiceDiscovery_test, ServiceDiscoveryIsNotifiedbyWaitSetAboutSingleService)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "f1cf36b5-3db2-4e6f-8e05-e7e449530ec0");
+    iox::popo::WaitSet<1U> waitSet;
+
+    waitSet
+        .attachEvent(sut,
+                     ServiceDiscoveryEvent::SERVICE_REGISTRY_HAS_CHANGED,
+                     0U,
+                     iox::popo::createNotificationCallback(testCallback))
+        .or_else([](auto) { GTEST_FAIL() << "Could not attach to wait set"; });
+
+    const iox::capro::ServiceDescription SERVICE_DESCRIPTION("Moep", "Fluepp", "Shoezzel");
+    iox::popo::UntypedPublisher publisher(SERVICE_DESCRIPTION);
+
+    auto notificationVector = waitSet.wait();
+
+    for (auto& notification : notificationVector)
     {
-        std::string instance = "i" + iox::cxx::convert::toString(i);
-        iox::capro::ServiceDescription SERVICE_DESCRIPTION(
-            SERVICE, IdString_t(iox::cxx::TruncateToCapacity, instance), "foo");
-        publishers.emplace_back(SERVICE_DESCRIPTION);
+        (*notification)();
     }
 
-    auto serviceContainer = sut.findService(SERVICE, iox::capro::Wildcard, iox::capro::Wildcard);
+    EXPECT_TRUE(callbackWasCalled);
+}
+
+TEST_F(ServiceDiscovery_test, ServiceDiscoveryNotifiedbyWaitSetFindsSingleService)
```

Comment:
1. Please add tests for the maximum supported amount of services and service results. A tests where you have maximum amount of `radar`, `frontleft`, `event1` ... `eventMAX_SUPPORTED` services and the user searches with one wildcard and a user searches with only wildcards to get all of them.

2. Furthermore, we require wild card tests where some services are found and some are not.

3. Test where we have one/multiple/all services are coming up and do a stop offer and cannot be found anymore

4. Test where one/multiple/all services are performing an offer and we find more and more (with wildcards and without)

Reply:
I agree, we need more tests for these scenarios but not in this PR. We should only check here whether we can attach and are correctly notified in case of a registry change.

The missing tests concern `findService` itself which this PR does not change semantically (only the implementation) and some of the existing cases need refactoring as well as a check for completeness (e.g. test partial wildcard searches etc.). There is an issue for this https://github.com/eclipse-iceoryx/iceoryx/issues/1065 and I suggest using this issue to improve the `findService` tests.

There is no point in checking attach and `findService` in combination if both are tested separately and we just call `findService` in an (anonymous) function on notification.

Reply:
@MatthiasKillat I agree the only problem I have is that we may release this feature but we actually did not really test it. I am fine that this is done in #1065 but then this issue has to be fully implemented until 28th of February (code freeze).

FYI @mossmaurice @elBoberido @FerdinandSpitzschnueffler @budrus

Reply:
I looked into the issue #1065 and spoke with @FerdinandSpitzschnueffler and as it seems the points I mentioned above are not considered in this issue!

The tests should be added here since the feature was implemented exactly in this pull request and tests should be strongly coupled with features.

Otherwise the two points from the reviewer checklist of every PR do not make sense:
* Unit tests have been written for new behavior
* Tests follow the [best practice for testing

It is also harder and more time consuming for a reviewer to check such things in a follow up PR.

Reply:
@elfenpiff I agree with you. However, it seems some of the tests are already implemented:

1. `FindServiceReturnsMaxServices` (line 469)
2. `FindServiceWithServiceWildcardReturnsCorrectServices` (line 251)
   1. Just three not one/many/max 
3. `FindServiceDoesNotReturnNotOfferedServices` (line 403)
4. Not implemented yet

I agree with @MatthiasKillat that unit test cases should not be duplicate for polling and push-based approach. So I would go ahead an implement 4). Does that make sense to you?

Reply:
If maximum and wildcard tests are not part of https://github.com/eclipse-iceoryx/iceoryx/issues/1065 then they need to be added or a new issues created (I favor adding them).

Reply:
@mossmaurice I agree with implementing only 4) Thanks for looking up the test cases!